src/Compiler/Facilities/prim-lexing.fs

```diff
@@ -325,6 +325,9 @@ and [<Sealed>] internal LexBuffer<'Char>
         with get () = bufferAcceptAction
         and set v = bufferAcceptAction <- v
 
+    member val private SourceTextBuilder = System.Text.StringBuilder()
+    member lexbuf.SourceText = lexbuf.SourceTextBuilder.ToString() |> SourceText.ofString
+
```

Comment:
Consider either pooling of the StringBuilder instances or a starting buffer that would reflect average size of an F# source file (the buffer above uses 4096 for example)